### PR TITLE
Staking squid v18

### PR DESCRIPTION
**Pull request by marc-aurele-besner**
## Staking squid v18

- Add more fields to track accrued total over a relatable base (epoch)

[squid-api](https://autonomys-labs.squids.live/staking-squid/v/v18/graphql)
[Hasura](https://autonomys-labs.squids.live/staking-squid/v/v18/addons/hasura/console)

<br class="Apple-interchange-newline">

___

### **PR Type**
enhancement, bug fix

___

### **Description**
- Introduced new types and interfaces for handling domains, execution receipts, and fraud proofs.
- Enhanced event processing logic for epoch transitions, unlocking, and withdrawals.
- Updated operator and nominator status handling with new fields and actions.
- Improved cache structure and statistics calculation for operators and domains.
- Added new GraphQL schema fields and enums for better data representation.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>15 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>v6.ts</strong><dd><code>Add new types and interfaces for domain and execution handling</code></dd></summary>
<hr>

indexers/staking-squid/src/types/v6.ts

<li>Introduced new types and interfaces for domain and execution receipt <br>handling.<br> <li> Added structures for fraud proof and bundle validity.<br> <li> Defined types for various domain-related proofs and receipts.<br>

</details>

  </td>
  <td><a href="https://github.com/autonomys/astral/pull/816/files#diff-7efcac01652237b93cfb9d5c139810937aadeecd06e3b26add45ef14a94e044b">+890/-0</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>storage.ts</strong><dd><code>Extend storage types and interfaces for v6 domains</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

indexers/staking-squid/src/types/domains/storage.ts

<li>Added support for v6 storage types and interfaces.<br> <li> Introduced new storage interfaces for domain receipts and sudo calls.<br> <li> Implemented storage handling for frozen domains.<br>

</details>

  </td>
  <td><a href="https://github.com/autonomys/astral/pull/816/files#diff-9b351a6d692ddff52011604a177f9b167a91d0e4f5d41cc5000269901dc9885d">+136/-0</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>epoch.ts</strong><dd><code>Enhance epoch transition event processing and status handling</code></dd></summary>
<hr>

indexers/staking-squid/src/events/epoch.ts

<li>Enhanced epoch transition event processing with new fields.<br> <li> Updated operator and nominator status handling.<br> <li> Improved calculation of accumulated epoch values.<br>

</details>

  </td>
  <td><a href="https://github.com/autonomys/astral/pull/816/files#diff-6797ac4345b3476e9e974e2878d029c1cebbcc44f773a77c7151fe56b803141f">+78/-19</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>unlock.ts</strong><dd><code>Implement event handlers for unlocking operations</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

indexers/staking-squid/src/events/unlock.ts

<li>Added new event handlers for unlocking operations.<br> <li> Implemented logic for processing operator and nominator unlock events.<br> <li> Updated withdrawal status handling.<br>

</details>

  </td>
  <td><a href="https://github.com/autonomys/astral/pull/816/files#diff-817f9c766238bec4b9d2fdaff743f944e400f393e4f707e2be2e21e39c2c966b">+202/-0</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>operator.ts</strong><dd><code>Enhance operator registration and withdrawal processes</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

indexers/staking-squid/src/extrinsics/operator.ts

<li>Enhanced operator registration and deregistration processes.<br> <li> Added logic for handling deposits and withdrawals.<br> <li> Updated operator and nominator status and actions.<br>

</details>

  </td>
  <td><a href="https://github.com/autonomys/astral/pull/816/files#diff-8c3290ae61e38f2b59281d1656f22d8e6b8319e434c699d2d1db163b023ed103">+85/-27</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>stats.ts</strong><dd><code>Improve statistics calculation and tracking for operators and domains</code></dd></summary>
<hr>

indexers/staking-squid/src/storage/stats.ts

<li>Improved statistics calculation for operators and domains.<br> <li> Added tracking for all-time high values.<br> <li> Enhanced stats creation with additional fields.<br>

</details>

  </td>
  <td><a href="https://github.com/autonomys/astral/pull/816/files#diff-db17cb75287be8535752607a3a6d4088f57352954e954275afbe4535c6abbef3">+90/-11</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>cache.ts</strong><dd><code>Update cache structure and enhance loading mechanisms</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

indexers/staking-squid/src/utils/cache.ts

<li>Updated cache structure to include deposits and withdrawals.<br> <li> Added new internal key store types for tracking high values.<br> <li> Enhanced cache loading and saving mechanisms.<br>

</details>

  </td>
  <td><a href="https://github.com/autonomys/astral/pull/816/files#diff-d16bcf285660b0ecc13b2eec2101d34d4ec51337a88abdaa242b4d16b6ee940b">+28/-17</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>calls.ts</strong><dd><code>Add new call types for domain sudo and treasury operations</code></dd></summary>
<hr>

indexers/staking-squid/src/types/domains/calls.ts

<li>Added new call types for domain sudo operations.<br> <li> Implemented freeze and unfreeze domain call types.<br> <li> Introduced treasury fund transfer call type.<br>

</details>

  </td>
  <td><a href="https://github.com/autonomys/astral/pull/816/files#diff-b18e2f513bd3fd3a49672951ca6300bef40387271f6675d66060283eeeafbe78">+94/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>operator.ts</strong><dd><code>Enhance operator event handling and nomination logic</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

indexers/staking-squid/src/events/operator.ts

<li>Enhanced operator nomination and slashing event handling.<br> <li> Updated logic for deposit creation and status updates.<br> <li> Improved handling of nominator actions during operator events.<br>

</details>

  </td>
  <td><a href="https://github.com/autonomys/astral/pull/816/files#diff-46ddd5ca8be87ad40782c027e2c8456a0f696f9f2f79bdcd31045e27d7fcabac">+47/-13</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>events.ts</strong><dd><code>Add new event types for domain operations and unlocking</code>&nbsp; &nbsp; </dd></summary>
<hr>

indexers/staking-squid/src/types/domains/events.ts

<li>Added new event types for domain operations.<br> <li> Implemented events for domain freezing and unfreezing.<br> <li> Introduced events for unlocking and pruning operations.<br>

</details>

  </td>
  <td><a href="https://github.com/autonomys/astral/pull/816/files#diff-9e2f1fb02926aad4d25484f9798c87e423623b23a8bae2581acef514cf323b75">+75/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>withdraw.ts</strong><dd><code>Enhance withdrawal event processing and status handling</code>&nbsp; &nbsp; </dd></summary>
<hr>

indexers/staking-squid/src/events/withdraw.ts

<li>Enhanced withdrawal event processing with new logic.<br> <li> Updated status handling for withdrawals.<br> <li> Improved calculation of estimated amounts.<br>

</details>

  </td>
  <td><a href="https://github.com/autonomys/astral/pull/816/files#diff-d0f552bf099c6661139211f05136d979a33561ca32a1160e30ef4257ab59ad6a">+42/-8</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>operator.ts</strong><dd><code>Update operator creation and status tracking</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

indexers/staking-squid/src/storage/operator.ts

<li>Updated operator creation with new fields and actions.<br> <li> Enhanced operator status and pending action tracking.<br> <li> Improved initialization with additional properties.<br>

</details>

  </td>
  <td><a href="https://github.com/autonomys/astral/pull/816/files#diff-fb606593529db8218dd8aafcb6f08a4fe71b19184a75a1b11ace2d806d481f35">+18/-3</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>index.ts</strong><dd><code>Integrate unlock event handlers and update processing logic</code></dd></summary>
<hr>

indexers/staking-squid/src/events/index.ts

<li>Integrated new unlock event handlers into event processing.<br> <li> Updated event processing logic for various domain events.<br> <li> Improved event handling structure.<br>

</details>

  </td>
  <td><a href="https://github.com/autonomys/astral/pull/816/files#diff-5b521ae1f9153379dff58d484ca88783464a10c686fbf61d27988aab2037175f">+28/-5</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>nominator.ts</strong><dd><code>Update nominator creation and status tracking</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

indexers/staking-squid/src/storage/nominator.ts

<li>Updated nominator creation with new fields and actions.<br> <li> Enhanced status and pending action tracking for nominators.<br> <li> Improved initialization with additional properties.<br>

</details>

  </td>
  <td><a href="https://github.com/autonomys/astral/pull/816/files#diff-72161d87f0e90b52ca336e48e21affa4b36e417cb59929d1b8ea25a339772003">+20/-3</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>schema.graphql</strong><dd><code>Update GraphQL schema with new fields and enums</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

indexers/staking-squid/schema.graphql

<li>Added new fields to domain, account, and operator types.<br> <li> Introduced new enums for pending actions and statuses.<br> <li> Enhanced schema with additional properties for deposits and <br>withdrawals.<br>

</details>

  </td>
  <td><a href="https://github.com/autonomys/astral/pull/816/files#diff-fd1a09ce347b260d78a002fac9e29a3c5306730d30873e3476dc5953edf3dfe4">+75/-19</a>&nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Additional files (token-limit)</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>1724072039632-Data.js</strong><dd><code>...</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

indexers/staking-squid/db/migrations/1724072039632-Data.js

...

</details>

  </td>
  <td><a href="https://github.com/autonomys/astral/pull/816/files#diff-5513ce5523b0d3cc28e5cfb1fc59ae63d84baa17efe92528f1209904e93ac3c8">+15/-27</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>gemini3h.jsonl</strong><dd><code>...</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

indexers/staking-squid/gemini3h.jsonl

...

</details>

  </td>
  <td><a href="https://github.com/autonomys/astral/pull/816/files#diff-9509e4d34b054461d1837c8fc860b0527c55546a494d721f863146574ef6d145">+1/-6</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: indexers/staking-squid/src/events/epoch.ts [hunk 1]
```diff
@@ -273,8 +294,9 @@ export async function processEpochTransitionEvent(
     operatorsIds.includes(o.operatorId.toString())
   );
 
-  for (const deposit of allDeposits) {
+  for (const [index, deposit] of allDeposits.entries()) {
```
Comment (clostao) [inline indexers/staking-squid/src/events/epoch.ts:297]: Could be simplified to just `allDeposits.values()`?

File: indexers/staking-squid/src/events/operator.ts [hunk 2]
```diff
@@ -64,6 +64,7 @@ export function processOperatorNominatedEvent(
       nominatorId: nominator.id,
       amount,
       storageFeeDeposit,
+      totalAmount: amount + storageFeeDeposit,
```
Comment (clostao) [inline indexers/staking-squid/src/events/operator.ts:67]: I don't understand why totalAmount is calculated this way: 

Shouldn't be `totalAmount` be the sent amount in the extrinsic? Or what's the same, the actual deposit (`amount` in the code) be extrinsic `amount` arg minus `storageFeeDeposit`?

File: indexers/staking-squid/src/events/epoch.ts (hunk 1 above)
Comment (clostao) [inline indexers/staking-squid/src/events/epoch.ts:297]: Could be simplified to `allDeposits(.values())`, right?

File: indexers/staking-squid/src/extrinsics/operator.ts [hunk 3]
```diff
@@ -82,6 +81,7 @@ export function processRegisterOperator(
       nominatorId: nominator.id,
       amount: amountBigInt,
       storageFeeDeposit,
+      totalAmount: amountBigInt + storageFeeDeposit,
```
Comment (clostao) [inline indexers/staking-squid/src/extrinsics/operator.ts:84]: Same as for operator OperatorNominatedEvent processor

File: indexers/staking-squid/src/events/operator.ts (hunk 2 above)
Comment (clostao) [reply indexers/staking-squid/src/events/operator.ts:67]: [This is the code](https://github.com/autonomys/subspace/blob/main/crates/pallet-domains/src/bundle_storage_fund.rs#L142 ) that regulates this mechanism. I think that it separates the extrinsic `amount` in two parts.

File: indexers/staking-squid/src/events/epoch.ts (hunk 1 above)
Comment (marc-aurele-besner) [reply indexers/staking-squid/src/events/epoch.ts]: yeah indeed the changes was not needed, I had changed to access the index, but I reverted the code, except that line, thanks

File: indexers/staking-squid/src/events/operator.ts (hunk 2 above)
Comment (marc-aurele-besner) [reply indexers/staking-squid/src/events/operator.ts]: Very good point, indeed my logic was wrong!
Thanks

**Review by clostao — DISMISSED** [review]
looks great 🚀

**Pull request merged**